streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -192,101 +194,117 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareActiveTask
+            graphConstructor
         );
-
-        final RackAwareGraphConstructor<UUID> graphConstructor = new MinTrafficGraphConstructor<>();
-        final AssignmentGraph assignmentGraph = buildTaskGraph(clientIds, clientRacks, taskIds, taskIdsByProcess, topicPartitionsByTaskId,
-            crossRackTrafficCost, nonOverlapCost, hasReplica, isStandby, graphConstructor);
         return assignmentGraph.graph.totalCost();
     }
 
-    private static AssignmentGraph buildTaskGraph(final List<UUID> clientIds,
+    private static AssignmentGraph buildTaskGraph(final Map<UUID, KafkaStreamsAssignment> assignmentsByUuid,
                                                   final Map<UUID, Optional<String>> clientRacks,
                                                   final List<TaskId> taskIds,
-                                                  final Map<UUID, Set<TaskId>> previousTaskIdsByProcess,
+                                                  final List<UUID> clientList,
                                                   final Map<TaskId, Set<TaskTopicPartition>> topicPartitionsByTaskId,
                                                   final int crossRackTrafficCost,
                                                   final int nonOverlapCost,
                                                   final boolean hasReplica,
                                                   final boolean isStandby,
-                                                  final RackAwareGraphConstructor<UUID> graphConstructor) {
-        final Map<UUID, UUID> clientsUuidByUuid = clientIds.stream().collect(Collectors.toMap(id -> id, id -> id));
+                                                  final RackAwareGraphConstructor<KafkaStreamsAssignment> graphConstructor) {
+        // Intentionally passed in empty -- these are actually outputs of the graph
         final Map<TaskId, UUID> clientByTask = new HashMap<>();
         final Map<UUID, Integer> taskCountByClient = new HashMap<>();
+
         final Graph<Integer> graph = graphConstructor.constructTaskGraph(
-            clientIds,
+            clientList,
             taskIds,
-            clientsUuidByUuid,
+            assignmentsByUuid,
             clientByTask,
             taskCountByClient,
-            (processId, taskId) -> {
-                return previousTaskIdsByProcess.get(processId).contains(taskId);
-            },
+            (assignment, taskId) -> assignment.assignedTaskIds().contains(taskId),
             (taskId, processId, inCurrentAssignment, unused0, unused1, unused2) -> {
                 final int assignmentChangeCost = !inCurrentAssignment ? nonOverlapCost : 0;
                 final Optional<String> clientRack = clientRacks.get(processId);
-                final Set<TaskTopicPartition> topicPartitions = topicPartitionsByTaskId.get(taskId).stream().filter(tp -> {
-                    return isStandby ? tp.isChangelog() : true;
-                }).collect(Collectors.toSet());
```

Comment:
fyi I'm not changing the logic here, just filtering the topicPartitions earlier. Figured this would simplify things especially in the active task case where we don't want to do any filtering anyways